### PR TITLE
8315462: [REDO] runtime/NMT/SummarySanityCheck.java failed with "Total committed (MMMMMM) did not match the summarized committed (NNNNNN)"

### PR DESCRIPTION
**Pull request by afshin-zafari**
This is a PR for the [original PR](https://github.com/openjdk/jdk/pull/16237).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315462](https://bugs.openjdk.org/browse/JDK-8315462): [REDO] runtime/NMT/SummarySanityCheck.java failed with "Total committed (MMMMMM) did not match the summarized committed (NNNNNN)" (**Bug** - P4)

### Reviewers
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [b71a424d](https://git.openjdk.org/jdk/pull/16262/files/b71a424d152a0c7a9795d1ffa972d994aebe3cea)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16262/head:pull/16262` \
`$ git checkout pull/16262`

Update a local copy of the PR: \
`$ git checkout pull/16262` \
`$ git pull https://git.openjdk.org/jdk.git pull/16262/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16262`

View PR using the GUI difftool: \
`$ git pr show -t 16262`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16262.diff">https://git.openjdk.org/jdk/pull/16262.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16262#issuecomment-1770304147)

### CODE REVIEW

File: src/hotspot/share/nmt/mallocTracker.cpp [hunk 1]
```diff
@@ -59,6 +59,23 @@ void MemoryCounter::update_peak(size_t size, size_t cnt) {
   }
 }
 
+void MallocMemorySnapshot::copy_to(MallocMemorySnapshot* s) {
+  // Need to make sure that mtChunks don't get deallocated while the
+  // copy is going on, because their size is adjusted using this
+  // buffer in make_adjustment().
```
Comment (tstuefe) [inline src/hotspot/share/nmt/mallocTracker.cpp:65]: Since you are moving this, could you change the comment a bit to make clear it only refers to the ThreadCritical, not the rest of the function? E.g. "Need to lock to make sure...."

File: src/hotspot/share/nmt/mallocTracker.hpp [hunk 2]
```diff
@@ -176,16 +178,12 @@ class MallocMemorySnapshot : public ResourceObj {
...
-     ThreadCritical tc;
-     s->_all_mallocs = _all_mallocs;
-     for (int index = 0; index < mt_number_of_types; index ++) {
-       s->_malloc[index] = _malloc[index];
-     }
-   }
+  inline size_t thread_count() const {
+    MallocMemorySnapshot* s = const_cast<MallocMemorySnapshot*>(this);
+    return s->by_type(mtThreadStack)->malloc_count();
+  }
```
Comment (tstuefe) [inline src/hotspot/share/nmt/mallocTracker.hpp:184]: Why is this needed?

Comment (afshin-zafari) [reply src/hotspot/share/nmt/mallocTracker.hpp:184]: It comes after merge, added in JDK-8315362.

File: src/hotspot/share/nmt/mallocTracker.cpp (hunk 1 above)
Comment (afshin-zafari) [reply src/hotspot/share/nmt/mallocTracker.cpp:65]: Done.

File: src/hotspot/share/nmt/mallocTracker.hpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/share/nmt/mallocTracker.hpp:184]: Yes, JDK-8315362 threw this out. IIRC we removed the awkward mechanism of counting threads by number of thread stacks. Hence my question, why the re-introduction? Merge glitch?

Comment (afshin-zafari) [reply src/hotspot/share/nmt/mallocTracker.hpp:184]: It's a merge glitch.

**Review by gerard-ziemski — APPROVED** [review]
Looks nice and clean!

(still need to remove the merge introduced code as Thomas pointed out though)

Thank you for fixing this!

**Comment by afshin-zafari** [comment]
@tstuefe, any more comments on this?

File: src/hotspot/share/nmt/mallocTracker.hpp [hunk 3]
```diff
@@ -55,6 +55,8 @@ class MemoryCounter {
  public:
   MemoryCounter() : _count(0), _size(0), _peak_count(0), _peak_size(0) {}
 
+  inline void set_size_and_count(size_t size, size_t count) { _size = size; _count = count; }
```
Comment (tstuefe) [inline src/hotspot/share/nmt/mallocTracker.hpp:58]: It occurred to me that this invalidates the peak counter. You need to update that too, e.g.

```
void set_size_and_count(size_t size, size_t count) { 
  _size = size;
  _count = count;
  update_peak(size, count);
}
```

**Review by tstuefe — CHANGES_REQUESTED** [review]

**Review by tstuefe — APPROVED** [review]
Okay

**Comment by afshin-zafari** [comment]
Thanks for your review @tstuefe and @gerard-ziemski.

/integrate
